File: Drone-Swarm-2/leader_follower/follower.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy
from geometry_msgs.msg import TwistStamped, PoseStamped
from mavros_msgs.srv import CommandTOL, CommandBool, SetMode  # Added SetMode import
from mavros_msgs.msg import State
import time
import threading
import sys

# Global lock to protect spin_until_future_complete calls
spin_lock = threading.Lock()

class FollowerController(Node):
    def __init__(self, drone_ns, leader_ns):
        super().__init__(f'{drone_ns}_controller')
        self.drone_ns = drone_ns
        self.leader_ns = leader_ns
        self.qos_profile = QoSProfile(reliability=ReliabilityPolicy.BEST_EFFORT, depth=10)
        
        # Increase queue depth to 100 for higher update throughput.
        qos_profile = QoSProfile(reliability=ReliabilityPolicy.BEST_EFFORT, depth=10)
        
        # Follower clients and publishers
        self.takeoff_client = self.create_client(CommandTOL, f'mavros/{drone_ns}/cmd/takeoff')
        self.disarm_client = self.create_client(CommandBool, f'mavros/{drone_ns}/cmd/arming')
        self.mode_client = self.create_client(SetMode, f'mavros/{drone_ns}/set_mode')  # Mode switching client
        while not self.mode_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for set_mode service...')
        
        self.velocity_publisher = self.create_publisher(TwistStamped, f'mavros/{drone_ns}/setpoint_velocity/cmd_vel', qos_profile)
        self.position_publisher = self.create_publisher(PoseStamped, f'mavros/{drone_ns}/setpoint_position/local', qos_profile)
        self.state_subscriber = self.create_subscription(State, f'mavros/{drone_ns}/state', self.state_callback, qos_profile)
        self.altitude_subscriber = self.create_subscription(PoseStamped, f'mavros/{drone_ns}/local_position/pose', self.altitude_callback, qos_profile)
        
        self.current_altitude = 0.0
        self.leader_altitude = None       # Leader's current altitude
        self.leader_initial_altitude = None  # Leader's initial altitude
        self.leader_initial_x = None  # Leader's initial x position
        self.armed = False
        self.offboard_mode = False

        # Follower's initial pose and altitude.
        self.initial_pose = None
        self.initial_altitude = None
        self.initial_x = None  # Follower's initial x position

        # Wait for services to be available.
        while not self.takeoff_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for takeoff service...')
        while not self.disarm_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for disarm service...')
        
        self.leader_pose_subscriber = None  # Initialize subscriber as None

    def create_leader_subscription(self):
        # Create subscription to leader's position
        self.leader_pose_subscriber = self.create_subscription(
            PoseStamped,
            f'mavros/{self.leader_ns}/local_position/pose',
            self.leader_pose_callback,
            self.qos_profile
        )

    def destroy_leader_subscription(self):
        if self.leader_pose_subscriber:
            self.destroy_subscription(self.leader_pose_subscriber)
            self.leader_pose_subscriber = None

    def state_callback(self, msg):
        self.armed = msg.armed
        self.offboard_mode = (msg.mode == "OFFBOARD")

    def altitude_callback(self, msg):
        self.current_altitude = msg.pose.position.z
        #self.get_logger().info(f"Follower altitude callback: {self.current_altitude}")
        if self.initial_pose is None:
            self.initial_pose = msg.pose
            self.initial_altitude = msg.pose.position.z
            self.initial_x = msg.pose.position.x
            '''self.get_logger().info(
                f"Follower initial position captured: x={self.initial_pose.position.x}, "
                f"y={self.initial_pose.position.y}, z={self.initial_altitude}"
            )'''

    def leader_pose_callback(self, msg):
        # Save the leader's current altitude and x position.
        self.leader_altitude = msg.pose.position.z
        self.leader_x = msg.pose.position.x
        #self.get_logger().info(f"Received leader current altitude: {self.leader_altitude}, x: {self.leader_x}")
        # Capture leader's initial altitude and x position once.
        if self.leader_initial_altitude is None:
            self.leader_initial_altitude = msg.pose.position.z
            self.leader_initial_x = msg.pose.position.x
            #self.get_logger().info(f"Leader initial altitude captured: {self.leader_initial_altitude}, x: {self.leader_initial_x}")

    def takeoff(self, altitude=1.0):
        req = CommandTOL.Request()
        req.altitude = altitude
        future = self.takeoff_client.call_async(req)
        with spin_lock:
            rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().success:
            self.get_logger().info(f"Follower takeoff successful to {altitude} meters")
            return True
        else:
            self.get_logger().info("Follower takeoff failed")
            return False

    def publish_velocity(self, z_velocity):
        msg = TwistStamped()
        msg.twist.linear.z = z_velocity
        self.velocity_publisher.publish(msg)

    def publish_position(self):
        if self.initial_pose is None:
            self.get_logger().error("Follower initial pose not received. Exiting.")
            rclpy.shutdown()
            sys.exit(1)
        if self.leader_altitude is None or self.leader_initial_altitude is None:
            self.get_logger().error("Leader data not received yet. Waiting...")
            return
        
        computed_target_z = self.leader_altitude - self.ground_offset
        computed_target_x = self.leader_x - (self.leader_initial_x - self.initial_x)
        target_z = computed_target_z
        target_x = computed_target_x

        msg = PoseStamped()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.pose.position.x = target_x
        msg.pose.position.y = self.initial_pose.position.y
        msg.pose.position.z = target_z
        self.position_publisher.publish(msg)
        #self.get_logger().info(f"Publishing follower target position: x={msg.pose.position.x}, z={target_z}")

    def switch_to_stabilize_mode(self):
        """
        Switch the flight mode to STABILIZE to allow disarming.
        """
        self.get_logger().info("Switching to STABILIZE mode for disarm...")
        req = SetMode.Request()
        req.custom_mode = "STABILIZE"
        future = self.mode_client.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().mode_sent:
            self.get_logger().info("STABILIZE mode set successfully.")
        else:
            self.get_logger().error("Failed to set STABILIZE mode.")
        time.sleep(2)  # Wait for the mode switch to take effect

    def disarm(self):
        req = CommandBool.Request()
        req.value = False
        future = self.disarm_client.call_async(req)
        with spin_lock:
            rclpy.spin_until_future_complete(self, future)
        if future.result() is not None and future.result().success:
            self.get_logger().info("Follower drone disarmed successfully")
            return True
        else:
            self.get_logger().info("Follower drone disarm failed")
            return False   
        
    def execute_mission(self):
        # First get both initial poses
        self.create_leader_subscription()
        self.get_logger().info("Waiting for initial poses...")
        
        while self.initial_pose is None or self.leader_initial_altitude is None:
            rclpy.spin_once(self, timeout_sec=0.1)
            time.sleep(0.1)
            
        # Store initial ground offset
        self.ground_offset = self.leader_initial_altitude - self.initial_altitude
        self.x_offset = self.leader_initial_x - self.initial_x
        #self.get_logger().info(f"Initial altitude offset: {self.ground_offset}, Initial x offset: {self.x_offset}")
        
        # Unsubscribe from leader
        self.destroy_leader_subscription()
        self.leader_altitude = None  # Reset leader altitude
        self.leader_x = None  # Reset leader x position
        
        if self.takeoff(1.0):
            self.get_logger().info("Holding at 1m for stabilization...")
            time.sleep(10)  # Wait for 5 seconds after takeoff

            # Now resubscribe to leader's position for following
            self.create_leader_subscription()
            self.get_logger().info("Resubscribed to leader's position")

            self.get_logger().info("Starting to follow leader's position...")
            
            # Continuously follow the leader's position
            while rclpy.ok():
                rclpy.spin_once(self, timeout_sec=0.1)
                #self.get_logger().info(f"Follower current altitude: {self.current_altitude}")
                
                # Ensure leader's altitude data is received
                if self.leader_altitude is None or self.leader_x is None:
                    self.get_logger().error("Leader data not received yet. Waiting...")
                    continue
                
                self.publish_position()
                time.sleep(0.1)  # Adjust the sleep time as needed

                # Check if the follower has landed
                if self.current_altitude < self.initial_altitude + 0.1:  # Adjust threshold as needed
                    print("altitude before disarming: ", self.current_altitude)
                    self.get_logger().info("Follower has landed. Disarming...")
                    break

            self.get_logger().info("Follower mission complete!")
            # Switch to a stabilizing flight mode before disarming.
            self.switch_to_stabilize_mode()
            self.disarm()
        return self.drone_ns

def main(args=None):
    rclpy.init(args=args)
    # For the follower, the drone namespace is "uas_1_2" and the leader's namespace is "uas_1_1"
    drone = FollowerController("uas_1_2", "uas_1_1")
    try:
        drone.execute_mission()
    except KeyboardInterrupt:
        pass
    drone.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: Drone-Swarm-2/cpp_files/follower.cpp
#include <chrono>
#include <memory>
#include <string>
#include <thread>
#include <mutex>
#include <cstdlib>
#include <deque>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist_stamped.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "mavros_msgs/srv/command_tol.hpp"
#include "mavros_msgs/srv/command_bool.hpp"
#include "mavros_msgs/srv/set_mode.hpp"
#include "mavros_msgs/msg/state.hpp"

using namespace std::chrono_literals;

std::mutex spin_mutex;

class FollowerController : public rclcpp::Node
{
public:
  FollowerController(const std::string & drone_ns, const std::string & leader_ns)
  : Node(drone_ns + "_controller"), drone_ns_(drone_ns), leader_ns_(leader_ns),
    current_altitude_(0.0), armed_(false), offboard_mode_(false),
    initial_altitude_(0.0), initial_pose_received_(false),
    leader_altitude_(0.0), leader_initial_altitude_(0.0),
    leader_initial_received_(false), leader_data_received_(false)
  {
    rclcpp::QoS qos_profile(rclcpp::KeepLast(10));
    qos_profile.best_effort();

    std::string base_ns = "mavros/" + drone_ns_;
    std::string leader_base = "mavros/" + leader_ns_;

    // Create follower service clients
    takeoff_client_ = this->create_client<mavros_msgs::srv::CommandTOL>(base_ns + "/cmd/takeoff");
    disarm_client_  = this->create_client<mavros_msgs::srv::CommandBool>(base_ns + "/cmd/arming");
    mode_client_    = this->create_client<mavros_msgs::srv::SetMode>(base_ns + "/set_mode");
    while (!mode_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "Waiting for set_mode service...");
    }

    // Create publishers
    velocity_publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(base_ns + "/setpoint_velocity/cmd_vel", qos_profile);
    position_publisher_ = this->create_publisher<geometry_msgs::msg::PoseStamped>(base_ns + "/setpoint_position/local", qos_profile);

    // Create subscriptions for follower data
    state_subscriber_ = this->create_subscription<mavros_msgs::msg::State>(
      base_ns + "/state", qos_profile,
      std::bind(&FollowerController::state_callback, this, std::placeholders::_1));
    altitude_subscriber_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      base_ns + "/local_position/pose", qos_profile,
      std::bind(&FollowerController::altitude_callback, this, std::placeholders::_1));

    // Subscribe to leader's pose
    leader_pose_subscriber_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      leader_base + "/local_position/pose", qos_profile,
      std::bind(&FollowerController::leader_pose_callback, this, std::placeholders::_1));

    while (!takeoff_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "Waiting for takeoff service...");
    }
    while (!disarm_client_->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "Waiting for disarm service...");
    }
  }

  bool takeoff(double altitude = 1.0)
  {
    auto req = std::make_shared<mavros_msgs::srv::CommandTOL::Request>();
    req->altitude = altitude;
    auto future = takeoff_client_->async_send_request(req);
    {
      std::lock_guard<std::mutex> lock(spin_mutex);
      rclcpp::spin_until_future_complete(this->get_node_base_interface(), future);
    }
    auto result = future.get();
    if (result != nullptr && result->success) {
      RCLCPP_INFO(this->get_logger(), "Follower takeoff successful to %.2f meters", altitude);
      return true;
    } else {
      RCLCPP_INFO(this->get_logger(), "Follower takeoff failed");
      return false;
    }
  }

  void publish_velocity(double z_velocity)
  {
    geometry_msgs::msg::TwistStamped msg;
    msg.twist.linear.z = z_velocity;
    velocity_publisher_->publish(msg);
  }

  void publish_position()
  {
    if (!initial_pose_received_) {
      RCLCPP_ERROR(this->get_logger(), "Follower initial pose not received. Exiting.");
      rclcpp::shutdown();
      std::exit(1);
    }
    if (!leader_initial_received_) {
      RCLCPP_ERROR(this->get_logger(), "Leader data not received yet. Waiting...");
      return;
    }
    double altitude_offset = leader_initial_altitude_ - initial_altitude_;
    double target_z = leader_altitude_ - altitude_offset;
    geometry_msgs::msg::PoseStamped msg;
    msg.header.stamp = this->get_clock()->now();
    msg.pose.position.x = initial_pose_.position.x;
    msg.pose.position.y = initial_pose_.position.y;
    msg.pose.position.z = target_z;
    position_publisher_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Publishing follower target altitude: %.2f", target_z);
  }

  void switch_to_stabilize_mode()
  {
    RCLCPP_INFO(this->get_logger(), "Switching to STABILIZE mode for disarm...");
    auto req = std::make_shared<mavros_msgs::srv::SetMode::Request>();
    req->custom_mode = "STABILIZE";
    auto future = mode_client_->async_send_request(req);
    rclcpp::spin_until_future_complete(this->get_node_base_interface(), future);
    auto result = future.get();
    if (result != nullptr && result->mode_sent) {
      RCLCPP_INFO(this->get_logger(), "STABILIZE mode set successfully.");
    } else {
      RCLCPP_ERROR(this->get_logger(), "Failed to set STABILIZE mode.");
    }
    std::this_thread::sleep_for(2s);
  }

  bool disarm()
  {
    auto req = std::make_shared<mavros_msgs::srv::CommandBool::Request>();
    req->value = false;
    auto future = disarm_client_->async_send_request(req);
    {
      std::lock_guard<std::mutex> lock(spin_mutex);
      rclcpp::spin_until_future_complete(this->get_node_base_interface(), future);
    }
    auto result = future.get();
    if (result != nullptr && result->success) {
      RCLCPP_INFO(this->get_logger(), "Follower drone disarmed successfully");
      return true;
    } else {
      RCLCPP_INFO(this->get_logger(), "Follower drone disarm failed");
      return false;
    }
  }

  std::string execute_mission()
  {
    while (!initial_pose_received_ || !leader_initial_received_) {
      rclcpp::spin_some(this->get_node_base_interface());
      RCLCPP_INFO(this->get_logger(), "Waiting for follower initial pose and leader initial altitude...");
      std::this_thread::sleep_for(20ms);
    }
    if (takeoff(1.0)) {
      std::this_thread::sleep_for(5s);
      double ascent_offset = 2.0;
      RCLCPP_INFO(this->get_logger(), "Follower starting mission based on leader altitude...");
      while (current_altitude_ < (initial_altitude_ + ascent_offset)) {
        rclcpp::spin_some(this->get_node_base_interface());
        RCLCPP_INFO(this->get_logger(), "Follower current altitude: %.2f", current_altitude_);
        publish_velocity(0.2);
        publish_position();
        std::this_thread::sleep_for(50ms);
      }
      RCLCPP_INFO(this->get_logger(), "Follower reached target altitude, starting descent...");
      while (current_altitude_ > (initial_altitude_ + 0.2)) {
        rclcpp::spin_some(this->get_node_base_interface());
        RCLCPP_INFO(this->get_logger(), "Follower current altitude: %.2f", current_altitude_);
        publish_velocity(-0.2);
        publish_position();
        std::this_thread::sleep_for(20ms);
      }
      publish_velocity(0.0);
      RCLCPP_INFO(this->get_logger(), "Follower mission complete!");
      switch_to_stabilize_mode();
      disarm();
    }
    return drone_ns_;
  }

private:
  void state_callback(const mavros_msgs::msg::State::SharedPtr msg)
  {
    armed_ = msg->armed;
    offboard_mode_ = (msg->mode == "OFFBOARD");
  }

  void altitude_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    double raw_altitude = msg->pose.position.z;
    altitude_history_.push_back(raw_altitude);
    if (altitude_history_.size() > filter_size_) {
      altitude_history_.pop_front();
    }

    double smoothed_altitude = std::accumulate(altitude_history_.begin(), altitude_history_.end(), 0.0) / altitude_history_.size();

    if (std::abs(smoothed_altitude - current_altitude_) > deadband_) {
      current_altitude_ = smoothed_altitude;
      RCLCPP_INFO(this->get_logger(), "Follower altitude callback: %.2f", current_altitude_);
    }

    if (!initial_pose_received_) {
      initial_pose_ = msg->pose;
      initial_altitude_ = msg->pose.position.z;
      initial_pose_received_ = true;
      RCLCPP_INFO(this->get_logger(), "Follower initial position captured: x=%.2f, y=%.2f, z=%.2f",
                  initial_pose_.position.x, initial_pose_.position.y, initial_altitude_);
    }
  }

  void leader_pose_callback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
  {
    leader_altitude_ = msg->pose.position.z;
    RCLCPP_INFO(this->get_logger(), "Received leader current altitude: %.2f", leader_altitude_);
    if (!leader_initial_received_) {
      leader_initial_altitude_ = msg->pose.position.z;
      leader_initial_received_ = true;
      RCLCPP_INFO(this->get_logger(), "Leader initial altitude captured: %.2f", leader_initial_altitude_);
    }
    leader_data_received_ = true;
  }

  // Members
  std::string drone_ns_;
  std::string leader_ns_;
  double current_altitude_;
  bool armed_;
  bool offboard_mode_;
  geometry_msgs::msg::Pose initial_pose_;
  bool initial_pose_received_;
  double initial_altitude_;

  double leader_altitude_;
  double leader_initial_altitude_;
  bool leader_initial_received_;
  bool leader_data_received_;

  std::deque<double> altitude_history_;
  const size_t filter_size_ = 10;
  const double deadband_ = 0.05;

  rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr velocity_publisher_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr position_publisher_;
  rclcpp::Subscription<mavros_msgs::msg::State>::SharedPtr state_subscriber_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr altitude_subscriber_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr leader_pose_subscriber_;
  rclcpp::Client<mavros_msgs::srv::CommandTOL>::SharedPtr takeoff_client_;
  rclcpp::Client<mavros_msgs::srv::CommandBool>::SharedPtr disarm_client_;
  rclcpp::Client<mavros_msgs::srv::SetMode>::SharedPtr mode_client_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<FollowerController>("uas_1_2", "uas_1_1");
  node->execute_mission();
  rclcpp::shutdown();
  return 0;
}
